File: AoC.2022/Day2.cs
using AoC.Library.Runner;
using AoC.Library.Utils;

namespace AoC._2022;

[DateInfo(2022, 2, AdventParts.All)]
public class Day2 : AdventSolution
{
    public override object SolvePartOne() => Input
        .Lines
        .Select(SolveLinePart1)
        .Sum();

    public override object SolvePartTwo()  => Input
        .Lines
        .Select(SolveLinePart2)
        .Sum();

    private int SolveLinePart1(string line)
    {
        var (opp, you) = line.SmartSplit().Select(Parse).ToArray().Unpack2();

        if (opp == you) return 3 + (int)you;
        if ((int)opp == ((int)you) % 3 + 1) return (int)you;

        return 6 + (int)you;
    }

    private int SolveLinePart2(string line)
    {
        var inps = line.SmartSplit().Select(Parse).ToArray();
        var opp = inps[0];
        var act = (Act)inps[1];

        var you = act switch {
            Act.Draw => opp,
            Act.Lose => opp switch {
                RPS.Rock => RPS.Scissors,
                RPS.Paper => RPS.Rock,
                _ => RPS.Paper
            },
            _ => opp switch {
                RPS.Rock => RPS.Paper,
                RPS.Paper => RPS.Scissors,
                _ => RPS.Rock
            }
        };

        WriteLine($"{opp} {you} {act}");
        return ((int)act - 1) * 3 + (int)(you);
    }

    private RPS Parse(string c) => c[0] switch {
        'A' or 'X' => RPS.Rock,
        'B' or 'Y' => RPS.Paper,
        _ => RPS.Scissors
    };
    
    private enum RPS
    {
        Rock = 1, Paper = 2, Scissors = 3
    }
    
    private enum Act
    {
        Lose = 1, Draw = 2, Win = 3
    }
    
}

File: AoC.2022.FSharp/Day2.fs
module AoC._2022.FSharp.Day2

open AoC.FSharp
open AoC.Library.Runner
open FSharpx


type private RPS =
    | Rock
    | Paper
    | Scissors

type private RPSResult =
    | Win
    | Lose
    | Draw

module private RPS =
    let parse =
        function
        | "X"
        | "A" -> Rock
        | "Y"
        | "B" -> Paper
        | "Z"
        | "C" -> Scissors
        | _ -> failwith "Invalid input"

    let parseResult =
        function
        | "X" -> Lose
        | "Y" -> Draw
        | "Z" -> Win
        | _ -> failwith "Invalid input"

    let compare a b =
        match a, b with
        | Rock, Scissors
        | Paper, Rock
        | Scissors, Paper -> Win
        | Rock, Paper
        | Paper, Scissors
        | Scissors, Rock -> Lose
        | _ -> Draw

    let find r a =
        match (a, r) with
        | Rock, Win -> Paper
        | Paper, Win -> Scissors
        | Scissors, Win -> Rock
        | Rock, Lose -> Scissors
        | Paper, Lose -> Rock
        | Scissors, Lose -> Paper
        | _, Draw -> a



    let toNumber =
        function
        | Rock -> 1
        | Paper -> 2
        | Scissors -> 3

    let resToNumber =
        function
        | Win -> 6
        | Draw -> 3
        | Lose -> 0

[<DateInfo(2022, 2, AdventParts.All)>]
type Day2() =
    inherit AdventSolution()

    override this.SolvePartOne() =
        this.Input.Lines
        |> Array.map (String.splitChar [| ' ' |] >> Array.map RPS.parse >> Tuple.ofArray2)
        |> Array.map (fun (a, b) -> ((RPS.compare b a) |> RPS.resToNumber) + (RPS.toNumber b))
        // |> Array.map (fun x ->
        //     printfn $"%d{x}"
        //     x)
        |> Array.sum
        :> obj


    override this.SolvePartTwo() =
        this.Input.Lines
        |> Array.map (String.splitChar [| ' ' |] >> Tuple.ofArray2)
        |> Array.map (fun (a, r) -> (RPS.parse a, RPS.parseResult r))
        |> Array.map (fun (a, r) -> (r |> RPS.resToNumber) + (a |> RPS.find r |> RPS.toNumber))
        |> Array.sum
        :> obj
